lightning/src/routing/router.rs

```diff
@@ -1564,45 +1564,61 @@ fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters,
 	for path in route.paths.iter_mut() {
 		let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
 
-		// Choose the last publicly known node as the starting point for the random walk
-		if let Some(starting_hop) = path.iter().rev().find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.pubkey))) {
-			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.pubkey);
+		// Remember the last three nodes of the random walk and avoid looping back on them.
+		// Init with the last three nodes from the actual path, if possible.
+		let last_id = NodeId::from_pubkey(&path.last().unwrap().pubkey);
```

Comment:
Out of curiosity, any reason why we don't check the entire path plus any selected shadow nodes? Just wondering if the additional complexity is justified if paths are effectively bounded.

Reply:
Ah, I see. I guess with this approach, though, you may still loop over one of the last three hops since they are overwritten by shadow hops. Is that a concern?

There may be a way to avoid that using iterators chained with three `Option`s (or an array of them). For example, with one shadow hop it would be something like:

```rust
let shadown_hop = None;
// ...
path.iter().rev().take(3).chain(shadow_hop.iter()).any(...)
```

Though you may have to do the pubkey to node id conversion more than once with a `map` somewhere in there.

Reply:
> super-duper-tiny-nit(cause I hope LLVM will optimize this out, though it may not): no need to create three `NodeId`s and then copy them into the array, you can just create them directly in the array initialization and put the array init on a few lines.

Ah, I assumed the compiler would optimize this away. Addressed with https://github.com/lightningdevkit/rust-lightning/pull/1360/commits/eb5a4c478aa2c98d0c04f38615e66296c673c20b.

> Ah, I see. I guess with this approach, though, you may still loop over one of the last three hops since they are overwritten by shadow hops. Is that a concern?

No, from my side it was intentional to just remember the last three hops and avoid having the random walk loop back on them. This should exclude many trivially implausible shadow routes, while still giving the random walk some degrees of freedom.
 
> There may be a way to avoid that using iterators chained with three `Option`s (or an array of them). For example, with one shadow hop it would be something like:
> 
> ```rust
> let shadown_hop = None;
> // ...
> path.iter().rev().take(3).chain(shadow_hop.iter()).any(...)
> ```
> 
> Though you may have to do the pubkey to node id conversion more than once with a `map` somewhere in there.

I'm not entirely sure what you are proposing, but I first played around with a similar construction to initialize `nodes_to_avoid`, i.e., something like `path.iter().rev().take(3).map(|h| NodeId::from_pubkey(&h.pubkey)`, which I however found to be not easily `collect`ed into an array.

Reply:
> > There may be a way to avoid that using iterators chained with three `Option`s (or an array of them). For example, with one shadow hop it would be something like:
> > ```rust
> > let shadown_hop = None;
> > // ...
> > path.iter().rev().take(3).chain(shadow_hop.iter()).any(...)
> > ```
...
> I'm not entirely sure what you are proposing,

The suggestion above was demonstrating how to avoid looping back on a single shadow hop and the original path's 3-suffix without allocations. It would require a little more work to support three shadow hops. And as note it would mean `NodeId::from_pubkey` would have to be called on each loop iteration. But given that `NodeId::from_pubkey` will allocate a `Vec`, maybe it's not worth it.

Reply:
Ah, after re-reading I'm getting what you wrote. Sorry, first time my brain failed to parse. Chaining iterators seems like a neat alternative way to go about this!

Reply:
It returns a fixed-length array, and passes the pointer to C, there's no allocation there - https://docs.rs/secp256k1/latest/src/secp256k1/key.rs.html#424-440

Reply:
Of course that said, `PublicKey::serialize` is *not* free, so if performance is the concern, we'd still want to avoid it, but its three calls so nbd.

Reply:
Ah, whoops. You're right. I was looking at the wrong type / implementation.

https://docs.rs/bitcoin/0.27.1/bitcoin/util/psbt/serialize/trait.Serialize.html